subprojects/build-cache/src/main/java/org/gradle/caching/internal/origin/OriginMetadataFactory.java

```diff
@@ -40,77 +35,65 @@
-    private static final String GRADLE_VERSION_KEY = "gradleVersion";
     private static final String CREATION_TIME_KEY = "creationTime";
     private static final String EXECUTION_TIME_KEY = "executionTime";
     private static final String ROOT_PATH_KEY = "rootPath";
     private static final String OPERATING_SYSTEM_KEY = "operatingSystem";
     private static final String HOST_NAME_KEY = "hostName";
     private static final String USER_NAME_KEY = "userName";
-    private static final Set<String> METADATA_KEYS = ImmutableSet.of(BUILD_INVOCATION_ID_KEY, TYPE_KEY, IDENTITY_KEY, GRADLE_VERSION_KEY, CREATION_TIME_KEY, EXECUTION_TIME_KEY, ROOT_PATH_KEY, OPERATING_SYSTEM_KEY, HOST_NAME_KEY, USER_NAME_KEY);
 
     private final InetAddressFactory inetAddressFactory;
     private final String userName;
     private final String operatingSystem;
-    private final Clock clock;
-    private final GradleVersion gradleVersion;
-    private final UniqueId currentBuildInvocationId;
+    private final String currentBuildInvocationId;
+    private final Consumer<Properties> additionalProperties;
     private final File rootDir;
 
-    public OriginMetadataFactory(Clock clock, InetAddressFactory inetAddressFactory, File rootDir, String userName, String operatingSystem, GradleVersion gradleVersion, UniqueId currentBuildInvocationId) {
+    public OriginMetadataFactory(InetAddressFactory inetAddressFactory, File rootDir, String userName, String operatingSystem, String currentBuildInvocationId, Consumer<Properties> additionalProperties) {
         this.inetAddressFactory = inetAddressFactory;
         this.rootDir = rootDir;
         this.userName = userName;
         this.operatingSystem = operatingSystem;
-        this.clock = clock;
-        this.gradleVersion = gradleVersion;
+        this.additionalProperties = additionalProperties;
         this.currentBuildInvocationId = currentBuildInvocationId;
     }
 
     public OriginWriter createWriter(CacheableEntity entry, long elapsedTime) {
         return new OriginWriter() {
             @Override
-            public void execute(OutputStream outputStream) {
-                // TODO: Replace this with something better
+            public void execute(OutputStream outputStream) throws IOException {
                 Properties properties = new Properties();
-                properties.setProperty(BUILD_INVOCATION_ID_KEY, currentBuildInvocationId.asString());
+                properties.setProperty(BUILD_INVOCATION_ID_KEY, currentBuildInvocationId);
                 properties.setProperty(TYPE_KEY, entry.getClass().getCanonicalName());
                 properties.setProperty(IDENTITY_KEY, entry.getIdentity());
-                properties.setProperty(GRADLE_VERSION_KEY, gradleVersion.getVersion());
-                properties.setProperty(CREATION_TIME_KEY, Long.toString(clock.getCurrentTime()));
+                properties.setProperty(CREATION_TIME_KEY, Long.toString(System.currentTimeMillis()));
```

Comment:
Using `Clock` sounds like a good idea to have a shared view of the time. Are you saying here that the creation time in the origin metadata is not important enough to warrant the use of `Clock`?

Reply:
IIUC, in the past we had some build scan events happening very close to each other end up with reversed timestamps, i.e. the event happening first would have the later timestamp. `Clock` is supposed to fix this and make time monotonic across threads.

In the origin metadata we don't really care about relative times. The timestamp in the metadata is not shown on the UI or used anywhere to calculate things. It's just a rough indication to the human looking inside the TAR.GZ file of when this artifact was created. For that purpose `System.currentTimeMillis()` should be more than sufficient.

Also, there's the scan data about the corresponding task (or whatever work item created the artifact) that includes all kinds of `Clock`-based times.